<h1>Ejercicio 3</h1>

## 1 Armar una tabla con

* todos los proyectos con nombres duplicados
* ordenados

### 1.1 Importando panda, numpy

In [1]:
import pandas
import numpy

### 1.2 Cargando datos de csv

In [2]:
file = "kickstarter-projects/ks-projects-201801.csv"
kickstarter_2018 = pandas.read_csv(file, parse_dates=["deadline","launched"], index_col=['ID'])

### 1.3 Tabla con proyectos duplicados

In [3]:
# Trayendo ids de proyectos duplicados
# keep false evita considerar los nombres en el vector ya que lo que necesitamos son los id solamente
ids_of_duplicated_projects = kickstarter_2018.name.duplicated(keep=False)

# Ordenando datos por el nombre del proyecto
kickstarter_2018[ids_of_duplicated_projects].sort_values(by='name')

name  \
ID                                                              
816998285   "...The Last shall be first..." LODB Lifestyle...   
815783250   "...The Last shall be first..." LODB Lifestyle...   
1010584633                                    "A Fresh Start"   
713417995                                     "A Fresh Start"   
1880084695      "American Sports Stories" - An Athletic Quest   
422509694       "American Sports Stories" - An Athletic Quest   
866225086                          "City of Mercy" (Canceled)   
1760892298                         "City of Mercy" (Canceled)   
1999958521                                "Color" Music Video   
870322805                                 "Color" Music Video   
1413519345  "Cut From the Illest Cloth" - The Brand New Fu...   
1637334549  "Cut From the Illest Cloth" - The Brand New Fu...   
1656736114                                              "ONE"   
159049492                                               "ONE"   
72198224                                        "On The Road"   
996180421                                       "On The Road"   
306461885   "Pulse"- a new album from "Blind Focus" to sup...   
1023301684  "Pulse"- a new album from "Blind Focus" to sup...   
516801004                                       "Reflections"   
2007967203                                      "Reflections"   
1223071711              "RunRainbow" Feature Film Sizzle Reel   
2136175444              "RunRainbow" Feature Film Sizzle Reel   
1165257902           "The Book of Sam" Book One Graphic Novel   
1606189204           "The Book of Sam" Book One Graphic Novel   
1823321664                            "What's in your spine?"   
465144515                             "What's in your spine?"   
234863316                                             #Elixir   
2043567899                                            #Elixir   
1889457470  $50 for a 1k word short story on a topic of yo...   
1165878498  $50 for a 1k word short story on a topic of yo...   
...                                                       ...   
255984219                                    noparkticket.com   
1945927325                                   noparkticket.com   
145035452                                  project (Canceled)   
1297288919                                 project (Canceled)   
1146277333               returning at a later date (Canceled)   
1556035715               returning at a later date (Canceled)   
391083153                returning at a later date (Canceled)   
1586289411               returning at a later date (Canceled)   
576148925                                         sFTP Client   
467374509                                         sFTP Client   
1331807604                                           smartLED   
216839969                                            smartLED   
167364797                                     test (Canceled)   
1404511608                                    test (Canceled)   
1117671370                                    test (Canceled)   
1492548024                    weSTAND: A Stand With a Mission   
1863478885                    weSTAND: A Stand With a Mission   
1334808206                                                  x   
774948469                                                   x   
1955983697                                       x (Canceled)   
983316237                                        x (Canceled)   
58345768                                         x (Canceled)   
1612055887                                       x (Canceled)   
234114061                                        x (Canceled)   
1839948714                                     xxx (Canceled)   
329580179                                      xxx (Canceled)   
1848699072                                                NaN   
634871725                                                 NaN   
648853978                                                 NaN   
796533179                  

## 2 Verificar

* proyectos que tienen nombres duplicados también tienen el hash de nombre duplicado

### 2.1 Importando md5 de hashlib

In [4]:
# Necesitamos esta libreria para codificar como hash los nombres de los proyectos
from hashlib import md5

### 2.2 Creando la función "convertir_a_hash"

In [5]:
def convertir_a_hash(val):
    if isinstance(val, float): 
        return str(val)
    return md5(val.encode('utf-8')).hexdigest()

### 2.3 Creando una nueva columna hashname con el nombre en hash

In [6]:
kickstarter_2018['hashname'] = kickstarter_2018['name'].apply(convertir_a_hash)

### 2.4 Verificando que el hash name también este duplicado

In [7]:
# Trayendo ids de proyectos duplicados
# keep false evita considerar los nombres en el vector ya que lo que necesitamos son los id solamente
ids_of_duplicated_projects = kickstarter_2018.name.duplicated(keep=False)

# Ordenando datos por el nombre del proyecto
kickstarter_2018[ids_of_duplicated_projects].sort_values(by='name')[['hashname','name']].head(5)

hashname  \
ID                                             
816998285   0c7a251ffe4c7834cbc4b04906952ff1   
815783250   0c7a251ffe4c7834cbc4b04906952ff1   
1010584633  67554ab4203d95f2f2f05365f768206e   
713417995   67554ab4203d95f2f2f05365f768206e   
1880084695  19a82bc4c5961834282575d07d9b5f7c   

                                                         name  
ID                                                             
816998285   "...The Last shall be first..." LODB Lifestyle...  
815783250   "...The Last shall be first..." LODB Lifestyle...  
1010584633                                    "A Fresh Start"  
713417995                                     "A Fresh Start"  
1880084695      "American Sports Stories" - An Athletic Quest

## 3 Comparar

* la cantidad de nombres raros en kickstarter_2018 con la que obtenemos al cargar 'ks-projects-201801.csv' con encoding iso-8859-1

### 3.1 Importando badness de ftfy

In [8]:
# Necesitamos esta libreria para establecer cuales son los valores raros
import ftfy.badness as badness

### 3.2 Creando la función es_raro 

In [9]:
def es_raro(val):
    if isinstance(val, float): 
        return 0
    return badness.sequence_weirdness(val)

### 3.3 Creando una columna 'nombres_raros' 

In [10]:
kickstarter_2018['nombres_raros'] = kickstarter_2018['name'].apply(es_raro)

### 3.4 Conteo de nombres raros

In [11]:
cantidad_de_nombres_raros = kickstarter_2018[kickstarter_2018['nombres_raros'] > 1].shape[0]
print ('Cantidad de nombres raros de kickstarter_2018 sin encoding', cantidad_de_nombres_raros)

Cantidad de nombres raros de kickstarter_2018 sin encoding 26


### 3.5 Conteo de nombres raros del dataset con encoding iso-8859-1

In [12]:
kickstarter_2018_con_encoding_iso_8859_1 = pandas.read_csv("kickstarter-projects/ks-projects-201801.csv", encoding='iso-8859-1')
kickstarter_2018_con_encoding_iso_8859_1['nombres_raros'] = kickstarter_2018_con_encoding_iso_8859_1['name'].apply(es_raro)

cantidad_de_nombres_raros_con_encoding_iso_8859_1 = kickstarter_2018_con_encoding_iso_8859_1[kickstarter_2018_con_encoding_iso_8859_1['nombres_raros'] > 1].shape[0]
print ('Cantidad de nombres raros en dataset con encoding iso 8859-1', cantidad_de_nombres_raros_con_encoding_iso_8859_1)




Cantidad de nombres raros en dataset con encoding iso 8859-1 7682


### 3.6 Analisis entre los dos datasets

* La cantidad de datos raros en el dataset de kickstarter sin encoding es de : 26
* La cantidad de datos raros en el dataset de kickstarted con encoding es de : 7682

## 4 Codificar

* Currency con dos estrategias

### 4.1 LabelEncoder

#### 4.1.1 Importando libreria preprocessing de sklearn

In [13]:
from sklearn import preprocessing

#### 4.1.2 Aplicando LabelEncoder sobre la columna "currency"

In [14]:
columna = 'currency'

file = "kickstarter-projects/ks-projects-201801.csv"
kickstarter_2018 = pandas.read_csv(file, parse_dates=["deadline","launched"], index_col=['ID'])

# Preparando columna currency para aplicar labelEncoder
labelEncoder = preprocessing.LabelEncoder()
labelEncoder.fit(kickstarter_2018[columna])

## se visualiza el mapa de encode
dict(zip(labelEncoder.classes_, labelEncoder.transform(labelEncoder.classes_)))

# Aplicando Label encoder sobre el campo currency
kickstarter_2018[columna] = labelEncoder.transform(kickstarter_2018[columna]) 
kickstarter_2018.head()

name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category  currency   deadline     goal            launched  \
ID                                                                           
1000002330    Publishing         5 2015-10-09   1000.0 2015-08-11 12:12:28   
1000003930  Film & Video        13 2017-11-01  30000.0 2017-09-02 04:43:57   
1000004038  Film & Video        13 2013-02-26  45000.0 2013-01-12 00:20:50   
1000007540         Music        13 2012-04-16   5000.0 2012-03-17 03:24:11   
1000011046  Film & Video        13 2015-08-29  19500.0 2015-07-04 08:35:03   

            pledged     state  backers country  usd pledged  usd_pledged_real  \
ID                                                                              
1000002330      0.0    failed        0      GB          0.0               0.0   
1000003930   2421.0    failed       15      US        100.0            2421.0   
1000004038    220.0    failed        3      US        220.0             220.0   
1000007540      1.0    failed        1      US          1.0               1.0   
1000011046   1283.0  canceled       14      US       1283.0            1283.0   

            usd_goal_real  
ID                         
1000002330        1533.95  
1000003930       30000.00  
1000004038       45000.00  
1000007540        5000.00  
1000011046       19500.00

#### 4.1.3 Aplicando la inversa de LabelEncoder sobre la columna "currency"

In [15]:
kickstarter_2018[columna] = labelEncoder.inverse_transform(kickstarter_2018[columna]) 
kickstarter_2018.head()

name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category currency   deadline     goal            launched  \
ID                                                                          
1000002330    Publishing      GBP 2015-10-09   1000.0 2015-08-11 12:12:28   
1000003930  Film & Video      USD 2017-11-01  30000.0 2017-09-02 04:43:57   
1000004038  Film & Video      USD 2013-02-26  45000.0 2013-01-12 00:20:50   
1000007540         Music      USD 2012-04-16   5000.0 2012-03-17 03:24:11   
1000011046  Film & Video      USD 2015-08-29  19500.0 2015-07-04 08:35:03   

            pledged     state  backers country  usd pledged  usd_pledged_real  \
ID                                                                              
1000002330      0.0    failed        0      GB          0.0               0.0   
1000003930   2421.0    failed       15      US        100.0            2421.0   
1000004038    220.0    failed        3      US        220.0             220.0   
1000007540      1.0    failed        1      US          1.0               1.0   
1000011046   1283.0  canceled       14      US       1283.0            1283.0   

            usd_goal_real  
ID                         
1000002330        1533.95  
1000003930       30000.00  
1000004038       45000.00  
1000007540        5000.00  
1000011046       19500.00

### 4.2 LabelBinarizer

#### 4.2.1 Importando libreria LabelBinarizer de preprocessing

In [18]:
# Ahora se aplica el encoding con LabelBinarizer
from sklearn.preprocessing import LabelBinarizer


#### 4.2.2 Aplicando LabelBinarizer sobre la columna "currency"

In [23]:
columna = 'currency'
labelBinarizer = LabelBinarizer()
labelBinarizer_result = labelBinarizer.fit_transform(kickstarter_2018[columna])
pandas.DataFrame(labelBinarizer_result, columns=((columna + '_') + pandas.Series(labelBinarizer.classes_))).head(10)

currency_AUD  currency_CAD  currency_CHF  currency_DKK  currency_EUR  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   
5             0             0             0             0             0   
6             0             0             0             0             0   
7             0             0             0             0             0   
8             0             0             0             0             0   
9             0             0             0             0             0   

   currency_GBP  currency_HKD  currency_JPY  currency_MXN  currency_NOK  \
0             1             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   
5             0             0             0             0             0   
6             0             0             0             0             0   
7             0             0             0             0             0   
8             0             0             0             0             0   
9             0             0             0             0             0   

   currency_NZD  currency_SEK  currency_SGD  currency_USD  
0             0             0             0             0  
1             0             0             0             1  
2             0             0             0             1  
3             0             0             0             1  
4             0             0             0             1  
5             0             0             0             1  
6             0             0             0             1  
7             0             0             0             1  
8             0             0             0             1  
9             0             0             0             1

## 5 